core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AdaptiveTermination.java

```diff
@@ -0,0 +1,171 @@
+package ai.timefold.solver.core.impl.solver.termination;
+
+import java.util.NavigableMap;
+import java.util.TreeMap;
+
+import ai.timefold.solver.core.api.score.Score;
+import ai.timefold.solver.core.impl.phase.scope.AbstractPhaseScope;
+import ai.timefold.solver.core.impl.phase.scope.AbstractStepScope;
+import ai.timefold.solver.core.impl.solver.scope.SolverScope;
+import ai.timefold.solver.core.impl.solver.thread.ChildThreadType;
+
+import org.jspecify.annotations.NonNull;
+import org.jspecify.annotations.Nullable;
```

Comment:
You can use `@NullMarked` which means that everything is `@NonNull` unless marked as `@Nullable`. In my experience, this saves a lot of hassle.